cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -136,14 +153,52 @@ protected void addPartition(PartitionKey key, String path, boolean addToExplore)
       put.add(Bytes.add(FIELD_PREFIX, Bytes.toBytes(entry.getKey())), // "f.<field name>"
               Bytes.toBytes(entry.getValue().toString()));            // "<string rep. of value>"
     }
+
+    addMetadataToPut(metadata, put);
+
     partitionsTable.put(put);
     partitionsAddedInSameTx.put(path, key);
-    if (addToExplore) {
+
+    if (explorable) {
       addPartitionToExplore(key, path);
       // TODO: make DDL operations transactional [CDAP-1393]
     }
   }
 
+  @Override
+  public void addMetadata(PartitionKey key, String metadataKey, String metadataValue) {
+    addMetadata(key, ImmutableMap.of(metadataKey, metadataValue));
+  }
+
+  @Override
+  public void addMetadata(PartitionKey key, Map<String, String> metadata) {
+    final byte[] rowKey = generateRowKey(key, partitioning);
+    Row row = partitionsTable.get(rowKey);
+    if (row == null) {
```

Comment:
Right, I either missed that, or just followed the pattern in getPartition method, which also checked for `row == null`.
I changed both of these places and added a relevant test case.